keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md

```diff
@@ -0,0 +1,1042 @@
++                       class := staticBinding.StorageClassName()
++                       storageResource := staticBinding.StorageResource()
++                       if _, ok := classResources[class]; !ok {
++                               classResources[class] = &StorageResource{
++                                       Requested: 0,
++                                       Capacity:  0,
++                               }
++                       }
++                       classResources[class].Requested += storageResource.Requested
++                       classResources[class].Capacity += storageResource.Capacity
++               }
++       } else {
++               // group dynamic biding volumes by storage class
++               for _, provision := range podVolumes.DynamicProvisions {
++                       if provision.Capacity == nil {
++                               continue
++                       }
++                       class := *provision.PVC.Spec.StorageClassName
++                       if _, ok := classResources[class]; !ok {
++                               classResources[class] = &StorageResource{
++                                       Requested: 0,
++                                       Capacity:  0,
++                               }
+                        }
++                       requestedQty := provision.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
++                       classResources[class].Requested += requestedQty.Value()
++                       classResources[class].Capacity += provision.Capacity.Capacity.Value()
+                }
+-               classResources[class].Requested += storageResource.Requested
+-               classResources[class].Capacity += storageResource.Capacity
+        }
++
+        return pl.scorer(classResources), nil
+}
+```
+
+Users can select the scoring logic from the following options in `VolumeBindingargs`. The scoring logic is the same among all Pod + PVC(s).
+
+- (a) Prefer a node with the least allocatable.
+- (b) Prefer a node with the maximum allocatable.
+
+Considering the common scenario of local storage, we want to leave room for volume expansion after node allocation. The default setting is to prefer a node with the maximum allocatable.
+
+### Conditions for scoring static or dynamic provisioning
+
+About the `Score` function, the score will be calculated with the existing way (only static provisioning is taken into account) if at least one PVC was statically provisioned. Otherwise, the score will be calculated from dynamic provisioning.
+
+Implementation idea:
+
+```diff
+func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	...
+
++ 	if len(static) != 0 {
++ 		return static_score, nil;	// Same value as the current method
++ 	} else {
++ 		return dynamic_score, nil;	// Propose in this KEP
++ 	}
```

Comment:
I'm confused by this, a couple questions:

1. shouldn't the scoring function itself be gated? 
2. a. why would we fallback to static scoring if there is a single PVC that was statically provisioned? 
   b. if there is a single statically provisioned PVC on each node in a cluster, wouldn't this mean that toggling dynamic scoring would effectively do nothing? That feels super confusing to me.

Reply:
Thank you for your feedback.

> 1. shouldn't the scoring function itself be gated?

Are you suggesting that the current `Score` function should be used for static provisioning, and a separate function should be implemented for dynamic provisioning?

> 2. a. why would we fallback to static scoring if there is a single PVC that was statically provisioned?
>    b. if there is a single statically provisioned PVC on each node in a cluster, wouldn't this mean that toggling dynamic scoring would effectively do nothing? That feels super confusing to me.

This KEP has gone through discussions and currently functions only when "dynamic provisioning is the only option available." Based on your feedback, it seems the discussion is leaning towards "combining the scores of static and dynamic provisioning to determine the final score." However, this discussion has [already taken place](https://github.com/kubernetes/enhancements/pull/4042#discussion_r1309463854), and the current design was chosen because combining the scores could lead to suboptimal choices due to score cancellation. Specifically, if the score for pre-created PVs is very high and the score for dynamic provisioning is low, adding them together could result in a suboptimal configuration([Ref](https://github.com/kubernetes/enhancements/pull/4042#discussion_r1323650156)). Therefore, the current design separates the scores for static and dynamic provisioning to avoid this issue.

Reply:
I believe what Han means is that the dynamic part of this check should only be checked when the feature gate is turned on, whereas when it's off it should only look at static. That's how I understand it, especially after reading through PRR question `###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?` below.